src/frontend/src/optimizer/plan_node/logical_multi_join.rs

```diff
@@ -0,0 +1,131 @@
+    pub base: PlanBase,
+    inputs: Vec<PlanRef>,
+    on: Condition,
+}
+
+impl fmt::Display for LogicalMultiJoin {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "LogicalMultiJoin {{ on: {} }}", &self.on)
+    }
+}
+
+impl LogicalMultiJoin {
+    pub(crate) fn new(base: PlanBase, inputs: Vec<PlanRef>, on: Condition) -> Self {
+        Self { base, inputs, on }
+    }
+
+    pub(crate) fn from_join(join: &PlanRef) -> Option<Self> {
+        let logical_join = join.as_logical_join()?;
+        if logical_join.join_type() != JoinType::Inner {
+            return None;
+        }
+        let left = logical_join.left();
+        let right = logical_join.right();
+
+        let left_col_num = left.schema().len();
+        let right_col_num = right.schema().len();
+
+        let mut inputs = vec![];
+        let mut conjunctions = logical_join.on().conjunctions.clone();
+        if let Some(multi_join) = left.as_logical_multi_join() {
+            inputs.extend(multi_join.inputs());
+            conjunctions.extend(multi_join.on().clone());
+        } else {
+            inputs.push(left.clone());
+        }
+        if let Some(multi_join) = right.as_logical_multi_join() {
+            inputs.extend(multi_join.inputs());
+            let right_on = multi_join.on().clone();
+            let mut mapping = ColIndexMapping::with_shift_offset(
+                left_col_num + right_col_num,
+                -(left_col_num as isize),
+            )
+            .inverse();
+            let new_on = right_on.rewrite_expr(&mut mapping);
+            conjunctions.extend(new_on.conjunctions);
+        } else {
+            inputs.push(right.clone());
+        }
+
+        Some(Self {
+            base: logical_join.base.clone(),
+            inputs,
+            on: Condition { conjunctions },
+        })
+    }
+
+    /// Get a reference to the logical join's on.
+    pub fn on(&self) -> &Condition {
+        &self.on
+    }
+
+    /// Clone with new `on` condition
+    pub fn clone_with_cond(&self, cond: Condition) -> Self {
+        Self::new(self.base.clone(), self.inputs.clone(), cond)
+    }
+}
+
+impl PlanTreeNode for LogicalMultiJoin {
+    fn inputs(&self) -> smallvec::SmallVec<[crate::optimizer::PlanRef; 2]> {
+        let mut vec = smallvec::SmallVec::new();
+        vec.extend(self.inputs.clone().into_iter());
+        vec
+    }
+
+    fn clone_with_inputs(&self, _inputs: &[crate::optimizer::PlanRef]) -> crate::optimizer::PlanRef {
+        todo!()
+    }
+}
+
+impl ToStream for LogicalMultiJoin {
+    fn logical_rewrite_for_stream(&self) -> Result<(PlanRef, ColIndexMapping)> {
+        todo!()
+    }
+
+    fn to_stream(&self) -> Result<PlanRef> {
+        todo!()
+    }
+}
+
+impl ToBatch for LogicalMultiJoin {
+    fn to_batch(&self) -> Result<PlanRef> {
+        todo!()
+    }
+}
+
+impl ColPrunable for LogicalMultiJoin {
+    fn prune_col(&self, _required_cols: &[usize]) -> PlanRef {
+        todo!()
```

Comment:
I think Todo should be replaced by panic!(), We never want to implement these methods as LogicalMultiJoin is just a placeholder node with a temporary lifetime.